### ISSUE TITLE
KPL aggregation does not guaranty order on consumer side

### ISSUE THREAD

**Issue by vzhabko** [awslabs/amazon-kinesis-producer#135]
When producer application sends records without delay between KPL addUserRecord() call, consumer receives records in mixed order, not in order records have been sent.
Code snippet to send records (2 shards in stream):

        for (int i = 0; i < 1000; i++) {
           if (id > 20) {
                id = 0;
            }
            record.setAccountId("Account-" + id); // used as a partition key
            record.setMessageId("message-" + i);
            record.setTime(System.currentTimeMillis());
            service.sendRecord(record); // converts to bytes and calls addUserRecord()
            id++;
        }

Fragment of Consumer output:
Message processed. Id:	message-894	accountId:	Account-12	shardId:	shardId-000000000000	Time taken:	1,884
Message processed. Id:	message-890	accountId:	Account-8	shardId:	shardId-000000000000	Time taken:	1,885
Message processed. Id:	message-900	accountId:	Account-18	shardId:	shardId-000000000000	Time taken:	1,884
Message processed. Id:	message-904	accountId:	Account-1	shardId:	shardId-000000000000	Time taken:	1,884
Message processed. Id:	message-892	accountId:	Account-10	shardId:	shardId-000000000000	Time taken:	1,885
Message processed. Id:	message-901	accountId:	Account-19	shardId:	shardId-000000000000	Time taken:	1,884

Mote: message Ids are not in order. If we put Thread.sleep(1) between calls to addUserRecord() all records will be consumed in correct order.
OS: Mac OS
KPL has default configuration: new KinesisProducer(new KinesisProducerConfiguration().setRegion(region));
KCL has default configuration as well. Uses:
import com.amazonaws.services.kinesis.clientlibrary.interfaces.v2.*

**Comment by davidh87** [comment]
I have also run into this, with aggregation enabled and disabled. The easiest observation is when aggregation is enabled: adding multiple user records for the same shard results in an aggregated record in the kinesis stream with some data records which are not in order. In particular, this seems to be more noticeable at the start of the data - of 100 records tested with, often there was an error in the first 16 (approx) but records after that seemed to be generally consistently correct. See https://pastebin.com/iGUpRh0b for a sample of that ended up in the stream - the ts fields of each record should be a series.

Like the original reporter, we've added a Thread.sleep(1) between the calls to addUserRecord which seems to have helped (although have observed some errors even with this set).

I'm also running on OS X. Am unsure if the bug is in the native libraries or the java library, since I cannot easily access the pipes being used to pass data between the two (so I can't easily observe where the our-of-order-ness is happening).

**Comment by davidh87** [comment]
In case others come here looking for fix:

Since ordering is so important to us, we've just migrated to using the kinesis SDK client and not the KPL. We're going to try adding aggregation by writing records in the format specified on https://github.com/awslabs/amazon-kinesis-producer/blob/master/aggregation-format.md with the hope that the KCL can still stream them as normal aggregated events.

**Comment by yotamshtosselfg** [comment]
It would really help if the records would be indeed published in a FIFO order.

Any traction on this?

**Comment by StephenTurley** [comment]
Yeah I don't understand how this isn't supported but its stated in the FAQ that maintaining order is a reason to choose Kinesis over SQS. 

> **When should I use Amazon Kinesis Data Streams, and when should I use Amazon SQS?**
> Ordering of records. For example, you want to transfer log data from the application host to the processing/archival host while maintaining the order of log statements.

https://aws.amazon.com/kinesis/data-streams/faqs/

**Comment by Sulejman26** [comment]
By default KPL has 24 connections, .setMaxConnections(1) on your KPL config should fix the problem.

**Comment by gabrielfmagalhaes** [comment]
> By default KPL has 24 connections, .setMaxConnections(1) on your KPL config should fix the problem.

So as I can see, you'll either get higher throughput or order between shards.
I'm also struggling with this issue, so I may change to implement the kinesis SDK client or think about using Lambdas for our scenario.

**Comment by Sulejman26** [comment]
> > By default KPL has 24 connections, .setMaxConnections(1) on your KPL config should fix the problem.
> 
> So as I can see, you'll either get higher throughput or order between shards. I'm also struggling with this issue, so I may change to implement the kinesis SDK client or think about using Lambdas for our scenario.

Yep, you can also benchmark with having a client for each shard or if you have lots of shards you can do some tests to decide how to split into multiple clients
